// react-rest-client/index.js
'use strict';

exports.__esModule = true;
exports.middleware = exports.withEndpoints = exports.withEndpoint = exports.Endpoints = exports.Endpoint = exports.Client = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Client = require('./Client');

var _Client2 = _interopRequireDefault(_Client);

var _Endpoint = require('./Endpoint');

var _Endpoint2 = _interopRequireDefault(_Endpoint);

var _Endpoints = require('./Endpoints');

var _Endpoints2 = _interopRequireDefault(_Endpoints);

var _middleware = require('./middleware');

var _middleware2 = _interopRequireDefault(_middleware);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var withEndpoint = function withEndpoint(props) {
  return function (Component) {
    return _react2.default.createElement(_Endpoint2.default, _extends({}, props, { component: Component }));
  };
};
var withEndpoints = function withEndpoints(props) {
  return function (Component) {
    return _react2.default.createElement(_Endpoints2.default, _extends({}, props, { component: Component }));
  };
};

exports.Client = _Client2.default;
exports.Endpoint = _Endpoint2.default;
exports.Endpoints = _Endpoints2.default;
exports.withEndpoint = withEndpoint;
exports.withEndpoints = withEndpoints;
exports.middleware = _middleware2.default;
